File: food_order_app/lib/domain/order/entities/order_entity.dart
import 'package:flutter/material.dart';

class OrderItemEntity {
  final String id;
  final String foodId;
  final String foodName;
  final String foodImageUrl;
  final double price;
  final int quantity;

  const OrderItemEntity({
    required this.id,
    required this.foodId,
    required this.foodName,
    required this.foodImageUrl,
    required this.price,
    required this.quantity,
  });

  double get totalPrice => price * quantity;

  Map<String, dynamic> toJson() => {
        'id': id,
        'foodId': foodId,
        'foodName': foodName,
        'foodImageUrl': foodImageUrl,
        'price': price,
        'quantity': quantity,
      };

  factory OrderItemEntity.fromJson(Map<String, dynamic> json) {
    return OrderItemEntity(
      id: json['id'] as String? ?? json['_id'] as String? ?? '',
      foodId: json['foodId'] as String? ?? '',
      foodName: json['foodName'] as String? ?? 'Food Item',
      foodImageUrl: json['foodImageUrl'] as String? ?? '',
      price: (json['price'] as num?)?.toDouble() ?? 0.0,
      quantity: (json['quantity'] as num?)?.toInt() ?? 1,
    );
  }
}

class OrderEntity {
  final String id;
  final String userId;
  final List<OrderItemEntity> items;
  final double totalAmount;
  final String deliveryAddress;
  final String status;
  final String paymentMethod;
  final String paymentStatus;
  final DateTime createdAt;
  final DateTime? updatedAt;
  final double? deliveryLat;
  final double? deliveryLng;
  final double? restaurantLat;
  final double? restaurantLng;

  const OrderEntity({
    required this.id,
    required this.userId,
    required this.items,
    required this.totalAmount,
    required this.deliveryAddress,
    required this.status,
    required this.paymentMethod,
    required this.paymentStatus,
    required this.createdAt,
    this.updatedAt,
    this.deliveryLat,
    this.deliveryLng,
    this.restaurantLat,
    this.restaurantLng,
  });

  factory OrderEntity.placeholder(String orderId) {
    return OrderEntity(
      id: orderId,
      userId: '',
      items: const [],
      totalAmount: 0.0,
      deliveryAddress: '',
      status: 'pending',
      paymentMethod: 'cash',
      paymentStatus: 'pending',
      createdAt: DateTime.now(),
    );
  }

  String get shortId {
    if (id.length >= 4) {
      final sub = id.substring(id.length - 4).toUpperCase();
      return '#BC-$sub';
    }
    return '#BC-$id';
  }

  bool get isActive =>
      status == 'pending' || status == 'preparing' || status == 'out_for_delivery';

  bool get isDelivered => status == 'delivered';

  bool get isCancelled => status == 'cancelled';

  String get statusLabel {
    switch (status) {
      case 'pending':
        return 'Order Placed';
      case 'preparing':
        return 'Kitchen Preparing';
      case 'out_for_delivery':
        return 'On the way';
      case 'delivered':
        return 'Delivered';
      case 'cancelled':
        return 'Cancelled';
      default:
        return status.toUpperCase();
    }
  }

  Color get statusColor {
    switch (status) {
      case 'pending':
        return const Color(0xFFF59E0B); // Amber
      case 'preparing':
        return const Color(0xFF3B82F6); // Blue
      case 'out_for_delivery':
        return const Color(0xFFF97316); // Brand Orange
      case 'delivered':
        return const Color(0xFF10B981); // Emerald Green
      case 'cancelled':
        return const Color(0xFFEF4444); // Red
      default:
        return const Color(0xFF6B7280);
    }
  }

  int get totalItemCount => items.fold(0, (sum, item) => sum + item.quantity);

  String get itemsSummary {
    if (items.isEmpty) return 'No items';
    return items.map((i) => '${i.quantity}x ${i.foodName}').join(', ');
  }

  String get formattedDate {
    final now = DateTime.now();
    final diff = now.difference(createdAt);

    final hour = createdAt.hour.toString().padLeft(2, '0');
    final minute = createdAt.minute.toString().padLeft(2, '0');

    if (diff.inDays == 0 && now.day == createdAt.day) {
      return 'Today, $hour:$minute';
    } else if (diff.inDays <= 1) {
      return 'Yesterday, $hour:$minute';
    } else {
      const months = [
        'Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun',
        'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'
      ];
      final month = months[createdAt.month - 1];
      return '${createdAt.day} $month ${createdAt.year}';
    }
  }
}

File: food_order_web/src/domain/orders/entities/order_entity.js
export const OrderStatus = Object.freeze({
  PENDING: 'pending',
  PREPARING: 'preparing',
  ON_THE_WAY: 'on_the_way',
  DELIVERED: 'delivered',
  CANCELLED: 'cancelled',
});

export class OrderItemEntity {
  constructor({
    id = '',
    foodId = '',
    foodName = 'Item',
    foodImageUrl = '',
    price = 0,
    quantity = 1,
  } = {}) {
    this.id = id;
    this.foodId = foodId;
    this.foodName = foodName;
    this.foodImageUrl = foodImageUrl;
    this.price = Number(price) || 0;
    this.quantity = Number(quantity) || 1;
  }

  get totalPrice() {
    return this.price * this.quantity;
  }
}

export class OrderEntity {
  constructor({
    id = '',
    orderNumber = '',
    userId = '',
    customerName = 'Guest Customer',
    customerPhone = '',
    items = [],
    totalAmount = 0,
    deliveryAddress = 'Phnom Penh, Cambodia',
    status = OrderStatus.PENDING,
    paymentMethod = 'cash',
    paymentStatus = 'pending',
    createdAt = new Date(),
    updatedAt = new Date(),
    notes = '',
  } = {}) {
    this.id = id;
    this.orderNumber = orderNumber || (id ? `#${id.slice(-6).toUpperCase()}` : '#ORD-0000');
    this.userId = userId;
    this.customerName = customerName;
    this.customerPhone = customerPhone;
    this.items = items.map((item) =>
      item instanceof OrderItemEntity ? item : new OrderItemEntity(item)
    );
    this.totalAmount = Number(totalAmount) || 0;
    this.deliveryAddress = deliveryAddress;
    this.status = (status || OrderStatus.PENDING).toLowerCase();
    this.paymentMethod = paymentMethod;
    this.paymentStatus = paymentStatus;
    this.createdAt = createdAt instanceof Date ? createdAt : new Date(createdAt);
    this.updatedAt = updatedAt instanceof Date ? updatedAt : new Date(updatedAt);
    this.notes = notes;
  }

  // Domain Business Getters
  get isPending() {
    return this.status === OrderStatus.PENDING;
  }

  get isPreparing() {
    return this.status === OrderStatus.PREPARING;
  }

  get isOnTheWay() {
    return this.status === OrderStatus.ON_THE_WAY;
  }

  get isDelivered() {
    return this.status === OrderStatus.DELIVERED;
  }

  get isCancelled() {
    return this.status === OrderStatus.CANCELLED;
  }

  get totalItemsCount() {
    return this.items.reduce((sum, item) => sum + item.quantity, 0);
  }

  get formattedTotalUSD() {
    return `$${this.totalAmount.toFixed(2)}`;
  }

  get formattedTotalKHR() {
    const khr = Math.round(this.totalAmount * 4100);
    return `${khr.toLocaleString()} ៛`;
  }

  get formattedDate() {
    try {
      return this.createdAt.toLocaleTimeString([], {
        hour: '2-digit',
        minute: '2-digit',
        hour12: true,
      });
    } catch {
      return '';
    }
  }

  get nextValidStatus() {
    switch (this.status) {
      case OrderStatus.PENDING:
        return OrderStatus.PREPARING;
      case OrderStatus.PREPARING:
        return OrderStatus.ON_THE_WAY;
      case OrderStatus.ON_THE_WAY:
        return OrderStatus.DELIVERED;
      default:
        return null;
    }
  }
}
